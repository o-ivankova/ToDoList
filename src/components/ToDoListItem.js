import React from 'react';

const ToDoListItem = ({ content, important = false }) => {
    const style = {
        color: important ? 'crimson' : 'black',
    }
    return <span style={ style }>{ content }</span>
}

export default ToDoListItem;
