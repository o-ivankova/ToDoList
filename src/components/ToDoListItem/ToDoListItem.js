import React from 'react';
import  './ToDoListItem.scss';
//import PropTypes from 'prop-types';

/* ToDoListItem.propTypes = {
  content: PropTypes.object,
}; */

// eslint-disable-next-line react/prop-types
const ToDoListItem = ({ content }) => {
  const [isDone, setIsDone] = React.useState(false);
  const [isImportant, setIsImportant] = React.useState(false);
 
  const onContentClick = () => { 
    setIsDone((prevState) => {
      return !prevState;
    })
  };

  const onMarkImportant = () => {
    setIsImportant((prevState) => {
      return !prevState
    });
  };
  
      return (
    <span span className={`todo-list-item ${isDone && ' done'} ${isImportant && ' important'}`}>
    <span className='todo-list-item-content' onClick={ onContentClick }>
        { content }
        </span>
        <button type="button" className="btn btn-outline-success btn-sm float-right" onClick={ onMarkImportant }>
            <i className='fa fa-exclamation'/>
        </button>

        <button type="button" className="btn btn-outline-danger btn-sm float-right">
        <i className="fa fa-trash-o" />
      </button>
        </span>
    );
}

export default ToDoListItem;
