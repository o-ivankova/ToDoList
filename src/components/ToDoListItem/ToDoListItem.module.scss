.todoListItem {
  width: 100%;
  margin: 0px;
  line-height: 45px;
  border-bottom-style: groove;
  border-width: 1px;
  border-radius: 10px;
  display: flex;
  justify-content: space-between;
  background-color: white;
  overflow-wrap: break-word;
}

.todoListItemCompleted {
  width: 100%;
  margin: 0px;
  line-height: 45px;
  border-bottom-style: groove;
  border-width: 1px;
  border-radius: 10px;
  display: flex;
  justify-content: space-between;
  background-color: white;
  overflow-wrap: break-word;
}

.todoListItemContent {
  user-select: none;
  vertical-align: middle;
  overflow-wrap: break-word;
}

.deleteBtn {
  border-style: hidden;
  border: 0px;
  background-color: inherit;
  vertical-align: middle;
  text-align: right;
  float: right;
  margin-left: 10px;
  margin-right: 20px;
}

.todoListItemCompleted .todoListItemContent {
  text-decoration: line-through;
}

.todoListItem .deleteBtn:hover,
.todoListItem .deleteBtn:focus{
  cursor: pointer;
 }