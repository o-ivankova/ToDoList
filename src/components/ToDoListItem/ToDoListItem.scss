.todo-list-item {
  width: 100%;
  margin: 0px;
  line-height: 45px;
  border-bottom-style: groove;
  border-width: 1px;
  display: flex;
  justify-content: space-between;
}

.todo-list-item .todo-list-item-content {
  cursor: pointer;
  user-select: none;
  vertical-align: middle;
}

.todo-list-item.completed .todo-list-item-content {
  text-decoration: line-through;
}

.todo-list-item .delete-btn {
  border-style: hidden;
  border: 0px;
  background-color: inherit;
  vertical-align: middle;
  text-align: right;
  float: right;
  margin-left: 10px;
  margin-right: 20px;
}

.todo-list-item .delete-btn:hover,
.todo-list-item .delete-btn:focus{
  cursor: pointer;
 }