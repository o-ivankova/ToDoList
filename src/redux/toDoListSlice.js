import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
