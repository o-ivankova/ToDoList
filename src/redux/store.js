import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);