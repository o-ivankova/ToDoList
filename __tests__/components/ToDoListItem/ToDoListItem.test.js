import React from "react";
import ToDoListItem from "../../../src/components/ToDoListItem/ToDoListItem";
import "@testing-library/jest-dom";
import { jest, expect, test, describe } from "@jest/globals";
import { fireEvent, render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import { setupStore } from "../../../src/redux/store";

describe("ToDoListItem tests", () => {
  const onDeletedMock = jest.fn();
  const onCheckCompletedMock = jest.fn();
  const content = "test";

  test("component is rendered", () => {
    render(
      <ToDoListItem
        content={content}
        completed={false}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    expect(screen.getByTestId("to-do-list-item")).not.toBeNull;
    expect(screen.getByTestId("todo-list-item-content")).not.toBeNull;
    expect(screen.getByTestId("checkbox")).not.toBeNull;
    expect(screen.getByTestId("delete-btn")).not.toBeNull;
  });

  test("the correct content is displayed", () => {
    render(
      <ToDoListItem
        content={content}
        completed={false}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    expect(screen.getByTestId("todo-list-item-content")).toHaveTextContent(
      content
    );
  });

  test("the correct checkbox value is displayed - unchecked", () => {
    render(
      <ToDoListItem
        content={content}
        completed={false}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    expect(screen.getByTestId("checkbox").checked).toBe(false);
  });

  test("the correct checkbox value is displayed - checked", () => {
    render(
      <ToDoListItem
        content={content}
        completed={true}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    expect(screen.getByTestId("checkbox").checked).toBe(true);
  });

  test("onDeleted function is called on delete button click", () => {
    render(
      <ToDoListItem
        content={content}
        completed={false}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    fireEvent.click(screen.getByTestId("delete-btn"));

    expect(onDeletedMock).toBeCalled;
  });

  test("checkbox click", () => {
    render(
      <ToDoListItem
        content={content}
        completed={false}
        onCheckCompleted={onCheckCompletedMock}
        onDeleted={onDeletedMock}
      />
    );

    fireEvent.click(screen.getByTestId("checkbox"));

    expect(onCheckCompletedMock).toBeCalled;
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
        elements: [],
      },
    };
    const mockStore = setupStore(preloadedState);
    const domTree = render(
      <Provider store={mockStore}>
        <ToDoListItem
          content={content}
          completed={false}
          onCheckCompleted={onCheckCompletedMock}
          onDeleted={onDeletedMock}
        />
      </Provider>
    );

    expect(domTree).toMatchSnapshot();
  });
});
