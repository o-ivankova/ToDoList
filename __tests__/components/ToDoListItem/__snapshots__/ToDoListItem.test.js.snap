// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoListItem tests snapshot test 1`] = `
<span
  className="todoListItem"
  data-testid="to-do-list-item"
>
  <div
    className="todoListItemContent"
    data-testid="todo-list-item-content"
  >
    <input
      checked={false}
      className="checkbox"
      data-testid="checkbox"
      onChange={
        [MockFunction] {
          "calls": Array [
            Array [
              SyntheticBaseEvent {
                "_reactName": "onChange",
                "_targetInst": null,
                "bubbles": true,
                "cancelable": true,
                "currentTarget": null,
                "defaultPrevented": false,
                "eventPhase": 3,
                "isDefaultPrevented": [Function],
                "isPropagationStopped": [Function],
                "isTrusted": false,
                "nativeEvent": MouseEvent {
                  "isTrusted": false,
                },
                "target": <input
                  class="checkbox"
                  data-testid="checkbox"
                  type="checkbox"
                />,
                "timeStamp": 1687445938519,
                "type": "change",
              },
            ],
          ],
          "results": Array [
            Object {
              "type": "return",
              "value": undefined,
            },
          ],
        }
      }
      type="checkbox"
    />
    test
  </div>
  <button
    className="deleteBtn"
    data-testid="delete-btn"
    onClick={
      [MockFunction] {
        "calls": Array [
          Array [
            SyntheticBaseEvent {
              "_reactName": "onClick",
              "_targetInst": null,
              "altKey": false,
              "bubbles": true,
              "button": 0,
              "buttons": 0,
              "cancelable": true,
              "clientX": 0,
              "clientY": 0,
              "ctrlKey": false,
              "currentTarget": null,
              "defaultPrevented": false,
              "detail": 0,
              "eventPhase": 3,
              "getModifierState": [Function],
              "isDefaultPrevented": [Function],
              "isPropagationStopped": [Function],
              "isTrusted": false,
              "metaKey": false,
              "movementX": 0,
              "movementY": 0,
              "nativeEvent": MouseEvent {
                "isTrusted": false,
              },
              "pageX": undefined,
              "pageY": undefined,
              "relatedTarget": null,
              "screenX": 0,
              "screenY": 0,
              "shiftKey": false,
              "target": <button
                class="deleteBtn"
                data-testid="delete-btn"
                type="button"
              >
                <svg
                  data-testid="delete-icon"
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 16 16"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708z"
                  />
                </svg>
              </button>,
              "timeStamp": 1687445938507,
              "type": "click",
              "view": null,
            },
          ],
        ],
        "results": Array [
          Object {
            "type": "return",
            "value": undefined,
          },
        ],
      }
    }
    type="button"
  >
    <svg
      data-testid="delete-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 16 16"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M4.646 4.646a.5.5 0 0 1 .708 0L8 7.293l2.646-2.647a.5.5 0 0 1 .708.708L8.707 8l2.647 2.646a.5.5 0 0 1-.708.708L8 8.707l-2.646 2.647a.5.5 0 0 1-.708-.708L7.293 8 4.646 5.354a.5.5 0 0 1 0-.708z"
      />
    </svg>
  </button>
</span>
`;
